src/electron/routing_service.ts

```diff
@@ -65,75 +63,79 @@ enum RoutingServiceStatusCode {
 //  - Linux: systemctl start|stop outline_proxy_controller.service
 //  - Windows: net start|stop OutlineService
 export class RoutingDaemon {
-  private socket: Socket|undefined;
+  private socket: Socket | undefined;
 
   private stopping = false;
 
   private fulfillDisconnect!: () => void;
 
-  private disconnected = new Promise<void>((F) => {
+  private disconnected = new Promise<void>(F => {
     this.fulfillDisconnect = F;
   });
 
   private networkChangeListener?: (status: TunnelStatus) => void;
 
   constructor(private proxyAddress: string, private isAutoConnect: boolean) {}
 
+  /**
+   * Make sure the routing daemon is installed in the system.
+   * A SystemConfigurationException will be thrown if installation failed.
+   */
+  public async ensureDaemonInstalled() {
+    if (await this.isDaemonInstalled()) {
```

Comment:
Let's tweak the design so that the layers are clean and we can properly do the UI.

This is my suggestion:
1. Make `VpnTunnel.connect` throw some "vpn not set up/configured"  error
1. Make the calling app code catch that and display the appropriate UI to the user:
  - On Windows, just tell them to rerun the installer.
  - On Linux
    1. use a dialog (it can be `window.confirm` for now, up to you) to tell the user that the vpn needs to be configured and needs permission.
    1. Trigger the installer with the permission prompt. The installer could be run by some `VpnTunnel.install/configure/setup()` method.
    1.  That setup can fail, in which case I guess we could suggest the user to reinstall the app or report feedback. We should at least output an install error.

Both the errors and dialogs need to use localized messages.

Reply:
So to clarify this, are you suggesting let's handle the error in the very top UI layer instead of the `RoutingDaemon` or `VpnTunnel`? We can have a discussion offline.